cryptographic_flows.md

```diff
@@ -0,0 +1,27 @@
+
+# Generate a secret
+
+This interaction is initiated by the local client application, when the user requests to create a new self-custodial secret.
+
+1. The calling application initiates generation of a new secret via the local client API. They specify the type of secret (e.g. arbitrary bytes, ECDSA key pair) and provide a communication session with the key server.
+
+2. The local client forwards the request to key server. Optionally, the local client augments the request with some random bytes.
+Possible failures: the session authentication fails or times out. The networking layer times out.
+
+2. The key server selects a unique ID (among its existing set of secret IDs) and assigns it to the secret.
+
+4. The key server runs the key generation protocol: the key server generates some random bytes and combines it with the random bytes provided by the local client to create the secret.
+If the secret is an ECDSA key pair, the key server may do additional processing on the secret to make sure it meets the requirements of an ECDSA private key and computes the corresponding public key. 
```

Comment:
Why don't we have local generation of the secret? There are many applications for which both local generation and caching of a secret would be appropriate, and this is more in line with our threat model. And remember that the current use case for an arbitrary secret is some action the _user_ takes, given the secret (as we are not talking about _distributed use_ yet). 

That is, our system provides "secure backup" in this case, rather than "secure use". We are building an e2e encrypted storage mechanism in this first phase of our development. The client should be sending an encrypted copy of their secret to the key server, and we should use OPAQUE to generate the encryption key, where this encryption key is known only to the client.

I think it would be helpful to clarify our design principles and threat model in a separate file, so that we can reference these more easily in the protocol specification; this should be written up in its own ticket. Using our internal design spec draft as an iteration start point would be good. However, it's worth clarifying a couple points here:
- We do not want the key server to know the secret, so the key server should not generate it. We do not fully trust the server and we want to use cryptographic protection wherever possible. This aligns with the design principle of _zero trust_.
- If we are worried about a lack of entropy on the local client device, then having the server augment the secret with randomness seems better than the other way around. This way, we have defense in depth and graceful failure: if the client has poor entropy sources, the key server may learn the secret via brute force attack, but the user is still protected from external adversaries. If the client has good entropy sources, then the key server will not be able to brute force the generated secret, and so the user is protected against an adversarial key server.